File: dev/_static/tutorials/user-interface/UserInterface.java
import krpc.client.Connection;
import krpc.client.RPCException;
import krpc.client.Stream;
import krpc.client.StreamException;
import krpc.client.services.SpaceCenter;
import krpc.client.services.SpaceCenter.Vessel;
import krpc.client.services.UI;
import krpc.client.services.UI.Button;
import krpc.client.services.UI.Canvas;
import krpc.client.services.UI.Layout;
import krpc.client.services.UI.Panel;
import krpc.client.services.UI.RectTransform;
import krpc.client.services.UI.Slider;
import krpc.client.services.UI.Text;

import org.javatuples.Pair;
import org.javatuples.Quartet;

import java.io.IOException;

public class UserInterface {
    public static void main(String[] args)
        throws IOException, RPCException, InterruptedException, StreamException {
        Connection connection = Connection.newInstance("User Interface Example");
        SpaceCenter spaceCenter = SpaceCenter.newInstance(connection);
        UI ui = UI.newInstance(connection);
        Canvas canvas = ui.getStockCanvas();

        // Get the size of the game window in pixels
        Pair<Double, Double> screenSize = canvas.getRectTransform().getSize();

        // Add a panel to contain the UI elements. The user can drag it around
        // the screen with the mouse.
        Panel panel = canvas.addPanel(true);
        panel.setDraggable(true);

        // Position the panel on the left of the screen
        RectTransform rect = panel.getRectTransform();
        rect.setSize(new Pair<Double,Double>(200.0, 120.0));
        rect.setPosition(
          new Pair<Double,Double>((110-(screenSize.getValue0())/2), 0.0));

        // Lay the contents of the panel out in a column
        Layout layout = panel.addVerticalLayout();
        layout.setPadding(new Quartet<Integer,Integer,Integer,Integer>(10, 10, 10, 10));
        layout.setSpacing(6);

        // Add a button to set the throttle to maximum, with a tooltip shown
        // while the mouse rests on it
        Button button = panel.addButton("Full Throttle", true);
        button.setTooltip("Set the throttle to maximum");
        button.getLayoutElement().setPreferredSize(new Pair<Double,Double>(-1.0, 30.0));

        // Add a slider to set the throttle by hand
        Slider slider = panel.addSlider(false, true);
        slider.getLayoutElement().setPreferredSize(new Pair<Double,Double>(-1.0, 20.0));

        // Add some text displaying the total engine thrust
        Text text = panel.addText("Thrust: 0 kN", true);
        text.setColor(new Quartet<Double,Double,Double,Double>(1.0, 1.0, 1.0, 1.0));
        text.setSize(18);

        // Set up streams to monitor the button and the slider
        Stream<Boolean> buttonClicked = connection.addStream(button, "getClicked");
        Stream<Boolean> sliderChanged = connection.addStream(slider, "getChanged");

        Vessel vessel = spaceCenter.getActiveVessel();
        while (true) {
            // Handle the throttle button being clicked
            if (buttonClicked.get ()) {
                vessel.getControl().setThrottle(1);
                slider.setValue(1);
                button.setClicked(false);
            }

            // Handle the user moving the throttle slider
            if (sliderChanged.get ()) {
                vessel.getControl().setThrottle(slider.getValue());
                slider.setChanged(false);
            }

            // Update the thrust text
            text.setContent(String.format("Thrust: %.0f kN", (vessel.getThrust()/1000)));

            Thread.sleep(1000);
        }
    }
}

File: dev/_static/tutorials/user-interface/UserInterface.lua
local krpc = require 'krpc'
local platform = require 'krpc.platform'
local List = require 'pl.List'
local conn = krpc.connect('User Interface Example')
local canvas = conn.ui.stock_canvas

-- Get the size of the game window in pixels
local screen_size = canvas.rect_transform.size

-- Add a panel to contain the UI elements. The user can drag it around the
-- screen with the mouse.
local panel = canvas:add_panel()
panel.draggable = true

-- Position the panel on the left of the screen
local rect = panel.rect_transform
rect.size = List{200, 120}
rect.position = List{110-(screen_size[1]/2), 0}

-- Lay the contents of the panel out in a column
local layout = panel:add_vertical_layout()
layout.padding = List{10, 10, 10, 10}
layout.spacing = 6

-- Add a button to set the throttle to maximum, with a tooltip shown while the
-- mouse rests on it
local button = panel:add_button("Full Throttle")
button.tooltip = "Set the throttle to maximum"
button.layout_element.preferred_size = List{-1, 30}

-- Add a slider to set the throttle by hand
local slider = panel:add_slider()
slider.layout_element.preferred_size = List{-1, 20}

-- Add some text displaying the total engine thrust
local text = panel:add_text("Thrust: 0 kN")
text.color = List{1, 1, 1, 1}
text.size = 18

local vessel = conn.space_center.active_vessel
while true do
    -- Handle the throttle button being clicked
    if button.clicked then
        vessel.control.throttle = 1
        slider.value = 1
        button.clicked = false
    end

    -- Handle the user moving the throttle slider
    if slider.changed then
        vessel.control.throttle = slider.value
        slider.changed = false
    end

    -- Update the thrust text
    text.content = string.format('Thrust: %.1f kN', vessel.thrust/1000)

    platform.sleep(0.1)
end
